File: Python/11.Container With Most Water.py
#在几个垂直于x轴的线段中，找出两个，使它俩与x轴围成的面积最大
class Solution(object):
    def maxArea(self, height):
        i,j,maxregion = 0,len(height)-1,0
        while i!=j:
            maxregion = max(maxregion,(j-i)*min(height[i],height[j]))
            if height[i]<height[j]:
                i += 1
            else:
                j -= 1
        return maxregion

s = Solution()
print(s.maxArea([3,4,10,8,9]))

File: C++/Array/11.Container With Most Water.cpp
/*
数组元素代表距离x轴的距离，这样形成一系列数组，求两条垂直线段和x轴能围成的最大容积。
*/
/*
思路：两个指针指头尾，更新最大容积，两个指针矮的向前挪。
 */
#include <iostream>
#include <vector>
#include <windows.h>
#include <windef.h>
#include <limits.h>
using namespace std;

class Solution {
public:
     int maxArea(vector<int>& height) {
        int i = 0, j = height.size()-1,maxarea = INT_MIN; 
        while(i < j)
        {
        	if(height[i]>=height[j])
        	{
        		maxarea = max(maxarea,(j-i)*height[j]);
        		--j;
        	}
        	else
        	{
        		maxarea = max(maxarea,(j-i)*height[i]);
        		++i;
        	}
        }
        return maxarea;
    }
};

int main()
{
	Solution s;
	vector<int> nums{5, 7, 2, 1};
	cout<<s.maxArea(nums)<<endl;
}

